### PR TITLE
BE-84 Resolve JSON structure in document result

### PR DESCRIPTION
**Pull request by MCFreddie777**

### CODE REVIEW

File: antiplag/migrations/0015_auto_20210413_2157.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            result["match_type"] = "CORPUS"
+
+        result["match_name"] = doc["name"]
+        result["ranges"] = doc["intervals"]
+        result["percentage"] = doc["percentage"]
+
+        return result
+
+    for result in Result.objects.all():
+        matched_docs = getattr(result, "matched_docs")
```
Comment (OliverKanat) [inline antiplag/migrations/0015_auto_20210413_2157.py:50]: Why are you not accessing `matches_docs` with dot notation?

File: antiplag/migrations/0015_auto_20210413_2157.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        for index, m_doc in enumerate(matched_docs):
+            Backup.objects.create(
+                document_id=getattr(result, "document_id"), **parse_matched_doc(m_doc)
+            )
```
Comment (OliverKanat) [inline antiplag/migrations/0015_auto_20210413_2157.py:55]: Why do you need to enumerate `matched_docs`? I don't see it used anywhere.

File: antiplag/migrations/0015_auto_20210413_2157.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        result["percentage"] = doc["percentage"]
+
+        return result
+
+    for result in Result.objects.all():
+        matched_docs = getattr(result, "matched_docs")
+
+        for index, m_doc in enumerate(matched_docs):
+            Backup.objects.create(
+                document_id=getattr(result, "document_id"), **parse_matched_doc(m_doc)
```
Comment (OliverKanat) [inline antiplag/migrations/0015_auto_20210413_2157.py:54]: Againg the `getattr`.

File: antiplag/migrations/0015_auto_20210413_2157.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    Backup = apps.get_model("antiplag", "result_backup")
+
+    def parse_matched_doc(doc):
+        """
+        Parses each matched doc and returns dict of values after migrate
+        """
+        result = {}
+        try:
+            # if is valid UUID, it's internal Document id
+            result["match_id"] = str(UUID(doc["id"], version=5))
```
Comment (OliverKanat) [inline antiplag/migrations/0015_auto_20210413_2157.py:36]: Is `version` arg necessary?

File: antiplag/migrations/0015_auto_20210413_2157.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            result["match_type"] = "UPLOADED"
```
Comment (OliverKanat) [inline antiplag/migrations/0015_auto_20210413_2157.py:37]: Use enumeration for `UPLOADED` and `CORPUS`.

File: antiplag/models.py [hunk 6]
```diff
@@ -41,42 +40,46 @@ class DocumentType(models.TextChoices):
...
 
-        merged_intervals = merge_intervals(indices)
+        return merge_intervals(indices)
 
-        return merged_intervals
+
+class Result(models.Model):
+    class MatchType(models.TextChoices):
+        CORPUS = "CORPUS"
+        UPLOADED = "UPLOADED"
```
Comment (OliverKanat) [inline antiplag/models.py:70]: Extract `MatchType` to `enums.py`.

Comment (OliverKanat) [reply antiplag/models.py:70]: I would update attribute values to shorter, same length literals. What do you think?

File: antiplag/models.py [hunk 7]
```diff
@@ -41,42 +40,46 @@ class DocumentType(models.TextChoices):
...
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    document = models.ForeignKey(
+        Document, on_delete=models.CASCADE, related_name="results"
+    )
+
```
Comment (OliverKanat) [inline antiplag/models.py:76]: Unnecessary new line.

File: antiplag/models.py [hunk 8]
```diff
@@ -41,42 +40,46 @@ class DocumentType(models.TextChoices):
...
+    match_id = models.CharField(max_length=255)
+    match_type = models.CharField(max_length=8, choices=MatchType.choices)
+    match_name = models.CharField(max_length=255)
+
```
Comment (OliverKanat) [inline antiplag/models.py:80]: Same here. IMHO it looks better grouped together and I haven't seen field attributes separated yet anywhere.

File: antiplag/serializers.py [hunk 9]
```diff
@@ -45,24 +45,28 @@ class Meta:
 
 
 class ResultSerializer(serializers.ModelSerializer):
+    id = serializers.CharField(source="match_id")
+    name = serializers.CharField(source="match_name")
+
     class Meta:
         model = Result
-        fields = ("percentage", "matched_docs")
+        fields = ("id", "name", "percentage")
```
Comment (OliverKanat) [inline antiplag/serializers.py:53]: Don't we need `ranges` from `matched_docs` to be serialized as well?

File: antiplag/tests/test_intervals.py [hunk 10]
```diff
@@ -6,6 +6,7 @@
 
 class ResultModelTestCase(TestCase):
     def setUp(self):
+        self.maxDiff = None
```
Comment (OliverKanat) [inline antiplag/tests/test_intervals.py:9]: Just curious, does it work on this project? I've tried it before, but with no luck.

**Review by OliverKanat — CHANGES_REQUESTED** [review]

File: antiplag/migrations/0015_auto_20210413_2157.py (hunk 1 above)
Comment (MCFreddie777) [reply antiplag/migrations/0015_auto_20210413_2157.py:50]: Not sure, I build upon kiko's function, therefore I used that. Will rework.

File: antiplag/models.py (hunk 6 above)
Comment (MCFreddie777) [reply antiplag/models.py:70]: Agreed on moving. About shorting it - IMHO it's more clear in the DB, and 5 characters don't make such difference. I would rather not.

File: antiplag/serializers.py (hunk 9 above)
Comment (MCFreddie777) [reply antiplag/serializers.py:53]: AFAIK, not at the moment. Actually, we use it - those ranges are are used in document property - those flattened and sorted, but we don't provide it to client.

File: antiplag/tests/test_intervals.py (hunk 10 above)
Comment (MCFreddie777) [reply antiplag/tests/test_intervals.py:9]: Yes, happened to me - when I had failing test, it showed up like FooLoremInput[234 more chars]EndingSequence.

When I used this option, the whole log was printed.

File: antiplag/migrations/0015_auto_20210413_2157.py (hunk 1 above)
Comment (OliverKanat) [reply antiplag/migrations/0015_auto_20210413_2157.py]: Kiko's case was different. It was generic function, hence the use of `getattr`.

File: antiplag/models.py (hunk 6 above)
Comment (OliverKanat) [reply antiplag/models.py]: OK, keep it as is.

**Review by OliverKanat — APPROVED** [review]

**Pull request merged**